### PR TITLE
[wip] Add codemods for moving from Cerebral 1.x to 2.x

### PR DESCRIPTION
**Pull request by NervosaX**
Addresses https://github.com/cerebral/cerebral/issues/318

### CODE REVIEW

File: codemods/1.x-2.x/__testfixtures__/controller.output.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import {Controller} from 'cerebral'
+
+const controller = Controller({
+  // You can add some initial state here if you want
+})
+
+export default controller
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/controller.output.js]: Should be:

```js
import {Controller} from 'cerebral'

const controller = Controller({
  state: {
    // You can add some initial state here if you want
  }
})

export default controller
```

File: codemods/1.x-2.x/__testfixtures__/operators-flow.input.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+} from 'cerebral/operators'
+
+export default [
+  // Conditional truthy check of state or input
+  when('state:foo.isAwesome'), {
+    true: [],
+    false: []
+  },
+
+  when('foo.isAwesome'), {
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/operators-flow.input.js]: I think this was not supported, it was `state:foo.isAwesome`...

File: codemods/1.x-2.x/__testfixtures__/operators-state.output.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import {
+  set,
+  copy,
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/operators-state.output.js]: No more `copy`... (not needed with operator tags)

File: codemods/1.x-2.x/__testfixtures__/operators-state.output.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  unset,
+  toggle,
+  state,
+  input,
+  output
+} from 'cerebral/operators'
+
+export default [
+  // Copy
+  set(state`foo.bar`, input`bar`),
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/operators-state.output.js]: This is

```js
set(input`bar`, state`foo.bar`)
```
```

File: codemods/1.x-2.x/__testfixtures__/operators-state.output.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  set(input`foo`, output`bar`),
+  set(input`foo`, state`foo.bar`),
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/operators-state.output.js]: Should be

```js
set(state`foo.bar`, input`foo`)
```

File: codemods/1.x-2.x/__testfixtures__/operators-state.output.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  set(state`foo.bar`, state`bar.baz`),
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/operators-state.output.js]: Should be

```js
set(state`bar.baz`, state`foo.bar`)
```

File: codemods/1.x-2.x/__testfixtures__/operators-state.output.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import {
+  set,
+  copy,
+  unset,
+  toggle,
+  state,
+  input,
+  output
```
Comment (henri-hulski) [inline codemods/1.x-2.x/__testfixtures__/operators-state.output.js]: Also no more `output`.

File: codemods/1.x-2.x/__testfixtures__/computed-decorator.output.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import {connect as Cerebral} from 'cerebral/react';
```
Comment (NervosaX) [inline codemods/1.x-2.x/__testfixtures__/computed-decorator.output.js]: https://github.com/cerebral/cerebral/blob/391a142da89628938c0c31263525d7aa084f3a3e/packages/demo/src/components/Tasks/index.js

Looking at this demo, I'd say it's still call connect.

```
@connect()
class ....
```

is the same as

```
connect(
  class ...
)
```

I just rename it to `@Cerebral` :)

File: codemods/1.x-2.x/__testfixtures__/operators-flow.output.js [hunk 9]
```diff
@@ -18,7 +14,7 @@ export default [
   },
 
   // Wait 200ms, then continue chain
-  ...wait(200, [
+  wait(200, [
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/operators-flow.output.js]: This is now `wait(200)` without any other parameter.

```js
wait(200),
doSomething,
```

Comment (NervosaX) [inline codemods/1.x-2.x/__testfixtures__/operators-flow.output.js]: Great, will fix it up.

File: codemods/1.x-2.x/__testfixtures__/operators-flow.output.js [hunk 10]
```diff
@@ -14,7 +14,5 @@ export default [
   },
 
   // Wait 200ms, then continue chain
-  wait(200, [
-    doSomething
-  ])
+  wait(200)
```
Comment (gaspard) [inline codemods/1.x-2.x/__testfixtures__/operators-flow.output.js]: But you need the `doSomething` here, after wait from how I understand the output...

Comment (NervosaX) [inline codemods/1.x-2.x/__testfixtures__/operators-flow.output.js]: Ah, crap. You're right.

Comment (NervosaX) [inline codemods/1.x-2.x/__testfixtures__/operators-flow.output.js]: This one is proving hard. Moving the array starts messing with the whole piece of code's whitespace.

**Comment by fvgoto** [comment]
If this PR is still worth it, I guess it should be updated according to
#660 rename `input` to `props`

**Comment by NervosaX** [comment]
Thanks @fvgoto. I think it's worth it, I am still waiting for some of these changes to stabilise. I'll definitely be using this to upgrade my codebase once 2.0 drops, so at least it's worth it for me. :)

**Comment by Guria** [comment]
ok, lets revisit this now with sequence/parallel, add some documentation and merge it in

**Comment by NervosaX** [comment]
Okay, I'll have a look into the new changes, and see if I can find nice ways to apply these mods to a codebase. 

We will need to document what it can't do automatically as well (for example the issues I had with the  delay->wait codemod that ruins code style).

There's currently nothing for converting from services.

**Comment by NervosaX** [comment]
@Guria Okay, I've fixed up the input tag -> props tag change.

I have yet to do the sequence/parallel change... that one might prove tricky to automate as well. Maybe we merge this in and work from master from here (others might feel free to contribute that way as well)

**Comment by Guria** [comment]
Did some squash and integrations with monorepo.
Merging it as is now.
Would be great to get usage instructions on website migration,

**Pull request merged**